openjdk/mmtkVMCompanionThread.cpp

```diff
@@ -123,23 +107,70 @@ void MMTkVMCompanionThread::wait_for_reached(stw_state desired_state) {
-  assert(Thread::current()->is_VM_thread(), "reach_suspended_and_wait_for_resume can only be executed by the VM thread");
+// Called by the VM thread in `VM_MMTkSTWOperation`.
+// This method notify that all Java threads have yielded, and will block the VM thread (thereby
+// blocking Java threads) until the GC requests start-the-world.
+void MMTkVMCompanionThread::do_mmtk_stw_operation() {
+  assert(Thread::current()->is_VM_thread(), "do_mmtk_stw_operation can only be executed by the VM thread");
 
-  MutexLockerEx locker(_lock, Mutex::_no_safepoint_check_flag);
+  {
+    MutexLockerEx locker(_lock, Mutex::_no_safepoint_check_flag);
 
-  // Tell the waiter thread that the world has stopped.
-  _reached_state = _threads_suspended;
-  _lock->notify_all();
+    // Tell the waiter thread that Java threads have stopped at yieldpoints.
+    _reached_state = _threads_suspended;
+    log_trace(gc)("do_mmtk_stw_operation: Reached _thread_suspended state. Notifying...");
+    _lock->notify_all();
+
+    // Wait until resume-the-world is requested
+    while (_desired_state != _threads_resumed) {
+      _lock->wait(Mutex::_no_safepoint_check_flag);
+    }
 
-  // Wait until resume-the-world is requested
-  while (_desired_state != _threads_resumed) {
-    _lock->wait(true);
+    // Tell the waiter thread that Java threads will eventually resume from yieldpoints.  This
+    // function will return, and, as soon as the VM thread stops executing safepoint VM operations,
+    // Java threads will resume from yieldpoints.
+    //
+    // Note: We have to notify *now* instead of after `VMThread::execute()`.  For reasons unknown
+    // (likely a bug in OpenJDK 11), the VMThread fails to notify the companion thread after
+    // evaluating `VM_MMTkSTWOperation`, and continues to execute other VM operations (such as
+    // `RevokeBias`). This leaves the companion thread blocking on `VMThread::execute()` until the
+    // VM thread finishes executing the next batch of queued VM operations.  If we notify after
+    // `VMThread::execute` in `run()`, it will cause a deadlock like the following:
+    //
+    // - The companion thread is blocked at `VMThread::execute()`, waiting for the next batch of VM
+    //   operations to finish.
+    // - The VM thread is blocked in `SafepointSynchronize::begin()`, waiting for all mutators to
+    //   reach safepoints.
+    // - One mutator is allocating too fast and triggers a GC, which requires the `WorkerMonitor`
+    //   lock in mmtk-core.
+    // - A GC worker is still executing `mmtk_resume_mutator`, holding the `WorkerMutator` (as the
+    //   last parked GC worker).  It is asking the companion thread to resume mutators, and is still
+    //   waiting for the companion thread to reach the `_thread_resumed` state.  As we see before,
+    //   the companion thread is waiting, too.
+    //
+    // By notifying now, we let the companion thread stop waiting, and therefore allowing the last
```

Comment:
I feel that this comment is confusing. If I understand correctly, the companion thread is not waiting on `_lock`, instead, it is waiting for the `VMThread` to notify it that the vm operation has finished. So the companion thread still needs to wait until the batch of vm operations to finish. However, because now gc threads no longer relies on the companion thread, they will be notified now deadlock no longer occurs.

 All other part looks good and makes perfect sense.

Reply:
You are right.  The comment is wrong.  The companion thread is not waiting on the `MMTkVMCompanionThread::_lock`, but on the `VMThread::VMOperationRequest_lock` while executing `VMThread::execute`.  If the VM thread notifies the `_lock` now, the companion thread will not stop waiting, but it will unblock the GC worker thread (running `resume_mutators`) which is waiting on `_lock` (in `MMTkVMCompanionThread::request`).

I'll rephrase the comment to "By notifying now, we unblock the GC worker waiting on `_lock` before the companion thread returns from `VMThread::execute`, and therefore allowing...".